presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -479,4 +516,15 @@ OutputBufferMemoryManager getMemoryManager()
     {
         return memoryManager;
     }
+
+    private void dereferencePage(SerializedPage pageSplit, Lifespan lifespan)
+    {
+        long outstandingPageCount = outstandingPageCountPerLifespan.get(lifespan).decrementAndGet();
```

Comment:
Is it possible that `setNoMorePagesForLifespan` and `dereferencePage` can race?
Something like, say, `dereferencePage` is called and `outstandingPageCount` became `0`, but say `noMorePagesForLifespan` does NOT contain the `lifespan` and we skip calling `lifespanCompletionCallback`, so this means the completion callback will not be called at all depending on the interleaving of these two methods.

Reply:
I was trying to analyze this.


THREAD 1: `setNoMorePagesForLifespan` -> `outstandingPageCount == 0` -> notify
THREAD 2: `decrementPageCount` -> `checkNoMorePagesForLifespan` -> notify

So it looks like there only 6 possible executions

1: 

`setNoMorePagesForLifespan`
`outstandingPageCount == 0`
`decrementPageCount` 
`checkNoMorePagesForLifespan`

2:

`decrementPageCount` 
`checkNoMorePagesForLifespan`
`setNoMorePagesForLifespan`
`outstandingPageCount == 0`

3: 

`decrementPageCount` 
`setNoMorePagesForLifespan`
`outstandingPageCount == 0`
`checkNoMorePagesForLifespan`

4:

`setNoMorePagesForLifespan`
`decrementPageCount` 
`checkNoMorePagesForLifespan`
`outstandingPageCount == 0`

5: 

`decrementPageCount` 
`setNoMorePagesForLifespan`
`checkNoMorePagesForLifespan`
`outstandingPageCount == 0`

6: 

`setNoMorePagesForLifespan`
`decrementPageCount` 
`outstandingPageCount == 0`
`checkNoMorePagesForLifespan`

In the last four the callback will be called twice, but that's not a problem.

I can't see an execution that could lead to a situation when the callback is not getting called.

Reply:
OK then I am probably missing something here. Is this a possible sequence:

```
T1: decrementPageCount
T1: outstandingPageCount == 0
T1: checkNoMorePagesForLifespan (Returns false)
(T1 doesn't call the callback)
T2: setNoMorePagesForLifespan
```

Reply:
Thanks @nezihyigitbasi  and @arhimondr  for reasoning into this. After T2 calls `setNoMorePagesForLifespan`, it's guaranteed to call `isFinishedForLifespan` and find it finished.